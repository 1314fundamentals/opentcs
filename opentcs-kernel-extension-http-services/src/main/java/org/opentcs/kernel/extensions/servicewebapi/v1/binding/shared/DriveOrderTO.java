// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi.v1.binding.shared;

import static java.util.Objects.requireNonNull;

import java.util.List;

/**
 */
public class DriveOrderTO {
  private String name;
  private DestinationTO destination;
  private String transportOrder;
  private RouteTO route;
  private StateTO state;

  public DriveOrderTO(
      String name,
      DestinationTO destination,
      String transportOrder,
      RouteTO route,
      StateTO state
  ) {
    this.name = requireNonNull(name, "name");
    this.destination = requireNonNull(destination, "destination");
    this.transportOrder = requireNonNull(transportOrder, "transportOrder");
    this.route = route;
    this.state = requireNonNull(state, "state");
  }

  public String getName() {
    return name;
  }

  public DriveOrderTO setName(String name) {
    this.name = requireNonNull(name, "name");
    return this;
  }

  public DestinationTO getDestination() {
    return destination;
  }

  public DriveOrderTO setDestination(DestinationTO destination) {
    this.destination = requireNonNull(destination, "destination");
    return this;
  }

  public String getTransportOrder() {
    return transportOrder;
  }

  public DriveOrderTO setTransportOrder(String transportOrder) {
    this.transportOrder = requireNonNull(transportOrder, "transportOrder");
    return this;
  }

  public RouteTO getRoute() {
    return route;
  }

  public DriveOrderTO setRoute(RouteTO route) {
    this.route = requireNonNull(route, "route");
    return this;
  }

  public StateTO getState() {
    return state;
  }

  public DriveOrderTO setState(StateTO state) {
    this.state = requireNonNull(state, "state");
    return this;
  }

  public static class DestinationTO {

    private String destination;
    private String operation;
    private List<Property> properties;

    public DestinationTO(
        String destination,
        String operation,
        List<Property> properties
    ) {
      this.destination = requireNonNull(destination, "destination");
      this.operation = requireNonNull(operation, "operation");
      this.properties = requireNonNull(properties, "properties");
    }

    public String getDestination() {
      return destination;
    }

    public DestinationTO setDestination(String destination) {
      this.destination = requireNonNull(destination, "destination");
      return this;
    }

    public String getOperation() {
      return operation;
    }

    public DestinationTO setOperation(String operation) {
      this.operation = requireNonNull(operation, "operation");
      return this;
    }

    public List<Property> getProperties() {
      return properties;
    }

    public DestinationTO setProperties(List<Property> properties) {
      this.properties = requireNonNull(properties, "properties");
      return this;
    }
  }

  // CHECKSTYLE:OFF
  public enum StateTO {
    PRISTINE,
    TRAVELLING,
    OPERATING,
    FINISHED,
    FAILED
  }
  // CHECKSTYLE:ON
}
