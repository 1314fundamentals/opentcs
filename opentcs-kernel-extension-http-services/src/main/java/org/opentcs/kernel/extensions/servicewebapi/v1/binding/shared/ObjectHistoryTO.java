// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi.v1.binding.shared;

import static java.util.Objects.requireNonNull;

import java.time.Instant;
import java.util.List;

/**
 */
public class ObjectHistoryTO {
  private List<ObjectHistoryEntryTO> entries;

  public ObjectHistoryTO(
      List<ObjectHistoryEntryTO> entries
  ) {
    this.entries = requireNonNull(entries, "entries");
  }

  public List<ObjectHistoryEntryTO> getEntries() {
    return entries;
  }

  public ObjectHistoryTO setEntries(List<ObjectHistoryEntryTO> entries) {
    this.entries = entries;
    return this;
  }

  /**
   */
  public static class ObjectHistoryEntryTO {
    private Instant timestamp;
    private String eventCode;
    private List<String> supplements;

    public ObjectHistoryEntryTO(
        Instant timeStamp,
        String eventCode,
        List<String> supplements
    ) {
      this.timestamp = requireNonNull(timeStamp, "timeStamp");
      this.eventCode = requireNonNull(eventCode, "eventCode");
      this.supplements = requireNonNull(supplements, "supplements");
    }

    public Instant getTimestamp() {
      return timestamp;
    }

    public ObjectHistoryEntryTO setTimestamp(Instant timestamp) {
      this.timestamp = timestamp;
      return this;
    }

    public String getEventCode() {
      return eventCode;
    }

    public ObjectHistoryEntryTO setEventCode(String eventCode) {
      this.eventCode = eventCode;
      return this;
    }

    public List<String> getSupplements() {
      return supplements;
    }

    public ObjectHistoryEntryTO setSupplements(List<String> supplements) {
      this.supplements = supplements;
      return this;
    }
  }
}
