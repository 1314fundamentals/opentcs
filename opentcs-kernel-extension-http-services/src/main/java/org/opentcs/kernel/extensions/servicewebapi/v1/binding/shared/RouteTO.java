// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi.v1.binding.shared;

import static java.util.Objects.requireNonNull;

import jakarta.annotation.Nonnull;
import jakarta.annotation.Nullable;
import java.util.List;

public class RouteTO {
  private long costs = -1;
  private List<Step> steps;

  public RouteTO(
      long costs,
      List<Step> steps
  ) {
    this.costs = costs;
    this.steps = requireNonNull(steps, "steps");
  }

  public long getCosts() {
    return costs;
  }

  public RouteTO setCosts(long costs) {
    this.costs = costs;
    return this;
  }

  public List<Step> getSteps() {
    return steps;
  }

  public RouteTO setSteps(List<Step> steps) {
    this.steps = steps;
    return this;
  }

  public static class Step {

    private String path;
    private String sourcePoint;
    private String destinationPoint;
    private VehicleOrientationTO vehicleOrientation;
    private int routeIndex;
    private long costs;
    private boolean executionAllowed;
    private ReroutingTypeTO reroutingType;

    public Step(
        String path,
        String sourcePoint,
        String destinationPoint,
        VehicleOrientationTO vehicleOrientation,
        int routeIndex,
        long costs,
        boolean executionAllowed,
        ReroutingTypeTO reroutingType
    ) {
      this.path = path;
      this.sourcePoint = sourcePoint;
      this.destinationPoint = requireNonNull(destinationPoint, "destinationPoint");
      this.vehicleOrientation = requireNonNull(vehicleOrientation, "vehicleOrientation");
      this.routeIndex = routeIndex;
      this.costs = costs;
      this.executionAllowed = executionAllowed;
      this.reroutingType = reroutingType;
    }

    @Nullable
    public String getPath() {
      return path;
    }

    public Step setPath(String path) {
      this.path = path;
      return this;
    }

    @Nullable
    public String getSourcePoint() {
      return sourcePoint;
    }

    public Step setSourcePoint(String sourcePoint) {
      this.sourcePoint = sourcePoint;
      return this;
    }

    @Nonnull
    public String getDestinationPoint() {
      return destinationPoint;
    }

    public Step setDestinationPoint(
        @Nonnull
        String destinationPoint
    ) {
      this.destinationPoint = requireNonNull(destinationPoint, "destinationPoint");
      return this;
    }

    @Nonnull
    public VehicleOrientationTO getVehicleOrientation() {
      return vehicleOrientation;
    }

    public Step setVehicleOrientation(
        @Nonnull
        VehicleOrientationTO vehicleOrientation
    ) {
      this.vehicleOrientation = requireNonNull(vehicleOrientation, "vehicleOrientation");
      return this;
    }

    public int getRouteIndex() {
      return routeIndex;
    }

    public Step setRouteIndex(int routeIndex) {
      this.routeIndex = routeIndex;
      return this;
    }

    public long getCosts() {
      return costs;
    }

    public Step setCosts(long costs) {
      this.costs = costs;
      return this;
    }

    public boolean isExecutionAllowed() {
      return executionAllowed;
    }

    public Step setExecutionAllowed(boolean executionAllowed) {
      this.executionAllowed = executionAllowed;
      return this;
    }

    public ReroutingTypeTO getReroutingType() {
      return reroutingType;
    }

    public Step setReroutingType(ReroutingTypeTO reroutingType) {
      this.reroutingType = reroutingType;
      return this;
    }

    // CHECKSTYLE:OFF
    public enum VehicleOrientationTO {
      FORWARD,
      BACKWARD,
      UNDEFINED
    }

    public enum ReroutingTypeTO {
      REGULAR,
      FORCED
    }
    // CHECKSTYLE:ON
  }
}
