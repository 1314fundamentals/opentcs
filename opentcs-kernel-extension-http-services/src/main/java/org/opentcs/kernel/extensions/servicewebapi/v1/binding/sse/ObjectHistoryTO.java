// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi.v1.binding.sse;

import java.time.Instant;
import java.util.List;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.experimental.Accessors;
import org.opentcs.data.ObjectHistory;

/**
 * A transfer object (to be used with the SSE API) representing an {@link ObjectHistory} instance.
 */
// CHECKSTYLE:OFF
@NoArgsConstructor
@Getter
@Setter
@Accessors(chain = true)
public class ObjectHistoryTO {

  private List<ObjectHistoryEntryTO> entries;

  @NoArgsConstructor
  @Getter
  @Setter
  @Accessors(chain = true)
  public static class ObjectHistoryEntryTO {

    private Instant timestamp;
    private String eventCode;
    private List<String> supplements;
  }
}
// CHECKSTYLE:ON
