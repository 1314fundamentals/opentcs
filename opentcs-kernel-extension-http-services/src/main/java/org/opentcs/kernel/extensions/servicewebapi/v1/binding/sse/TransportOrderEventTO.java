// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi.v1.binding.sse;

import java.time.Instant;
import java.util.List;
import java.util.Map;
import javax.annotation.Nullable;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.experimental.Accessors;
import org.opentcs.data.TCSObjectEvent;

/**
 * A transfer object (to be used with the SSE API) representing a {@link TCSObjectEvent} for a
 * transport order.
 */
public class TransportOrderEventTO {

  private final TransportOrderTO currentObjectState;
  private final TransportOrderTO previousObjectState;

  public TransportOrderEventTO(
      @Nullable
      TransportOrderTO currentObjectState,
      @Nullable
      TransportOrderTO previousObjectState
  ) {
    this.currentObjectState = currentObjectState;
    this.previousObjectState = previousObjectState;
  }

  public TransportOrderTO getCurrentObjectState() {
    return currentObjectState;
  }

  public TransportOrderTO getPreviousObjectState() {
    return previousObjectState;
  }

  // CHECKSTYLE:OFF
  @NoArgsConstructor
  @Getter
  @Setter
  @Accessors(chain = true)
  public static class TransportOrderTO {

    private String name;
    private Map<String, String> properties;
    private ObjectHistoryTO history;
    private String type;
    private List<String> dependencies;
    private List<DriveOrderTO> driveOrders;
    private String peripheralReservationToken;
    private int currentDriveOrderIndex;
    private int currentRouteStepIndex;
    private StateTO state;
    private Instant creationTime;
    private Instant deadline;
    private Instant finishedTime;
    private String intendedVehicle;
    private String processingVehicle;
    private String wrappingSequence;
    private boolean dispensable;
  }

  @NoArgsConstructor
  @Getter
  @Setter
  @Accessors(chain = true)
  public static class DriveOrderTO {

    private String name;
    private DestinationTO destination;
    private String transportOrder;
    private RouteTO route;
    private StateTO state;

    @NoArgsConstructor
    @Getter
    @Setter
    @Accessors(chain = true)
    public static class DestinationTO {

      private String destination;
      private String operation;
      private Map<String, String> properties;
    }

    @NoArgsConstructor
    @Getter
    @Setter
    @Accessors(chain = true)
    public static class RouteTO {

      private List<StepTO> steps;
      private long costs;

      @NoArgsConstructor
      @Getter
      @Setter
      @Accessors(chain = true)
      public static class StepTO {

        private String path;
        private String sourcePoint;
        private String destinationPoint;
        private VehicleOrientationTO vehicleOrientation;
        private int routeIndex;
        private long costs;
        private boolean executionAllowed;
        private ReroutingTypeTO reroutingType;
      }

      public enum VehicleOrientationTO {
        FORWARD,
        BACKWARD,
        UNDEFINED
      }

      public enum ReroutingTypeTO {
        REGULAR,
        FORCED
      }
    }

    public enum StateTO {
      PRISTINE,
      TRAVELLING,
      OPERATING,
      FINISHED,
      FAILED
    }
  }

  public enum StateTO {
    RAW,
    ACTIVE,
    DISPATCHABLE,
    BEING_PROCESSED,
    WITHDRAWN,
    FINISHED,
    FAILED,
    UNROUTABLE
  }
  // CHECKSTYLE:ON
}
