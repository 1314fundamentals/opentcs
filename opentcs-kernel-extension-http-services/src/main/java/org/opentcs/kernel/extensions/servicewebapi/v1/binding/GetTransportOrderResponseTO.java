// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi.v1.binding;

import static java.util.Objects.requireNonNull;

import jakarta.annotation.Nonnull;
import java.util.ArrayList;
import java.util.List;
import org.opentcs.data.order.TransportOrder;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.shared.DestinationState;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.shared.Property;

/**
 */
public class GetTransportOrderResponseTO {

  private boolean dispensable;

  private String name = "";

  private String peripheralReservationToken;

  private String wrappingSequence;

  private String type = "";

  private TransportOrder.State state = TransportOrder.State.RAW;

  private String intendedVehicle;

  private String processingVehicle;

  private List<DestinationState> destinations = new ArrayList<>();

  @Nonnull
  private List<Property> properties = List.of();

  public GetTransportOrderResponseTO() {
  }

  public boolean isDispensable() {
    return dispensable;
  }

  public GetTransportOrderResponseTO setDispensable(boolean dispensable) {
    this.dispensable = dispensable;
    return this;
  }

  public String getName() {
    return name;
  }

  public GetTransportOrderResponseTO setName(String name) {
    this.name = requireNonNull(name, "name");
    return this;
  }

  public String getPeripheralReservationToken() {
    return peripheralReservationToken;
  }

  public GetTransportOrderResponseTO setPeripheralReservationToken(
      String peripheralReservationToken
  ) {
    this.peripheralReservationToken = peripheralReservationToken;
    return this;
  }

  public String getWrappingSequence() {
    return wrappingSequence;
  }

  public GetTransportOrderResponseTO setWrappingSequence(String wrappingSequence) {
    this.wrappingSequence = wrappingSequence;
    return this;
  }

  public String getType() {
    return type;
  }

  public GetTransportOrderResponseTO setType(String type) {
    this.type = type;
    return this;
  }

  public TransportOrder.State getState() {
    return state;
  }

  public GetTransportOrderResponseTO setState(TransportOrder.State state) {
    this.state = requireNonNull(state, "state");
    return this;
  }

  public String getIntendedVehicle() {
    return intendedVehicle;
  }

  public GetTransportOrderResponseTO setIntendedVehicle(String intendedVehicle) {
    this.intendedVehicle = intendedVehicle;
    return this;
  }

  public String getProcessingVehicle() {
    return processingVehicle;
  }

  public GetTransportOrderResponseTO setProcessingVehicle(String processingVehicle) {
    this.processingVehicle = processingVehicle;
    return this;
  }

  public List<DestinationState> getDestinations() {
    return destinations;
  }

  public GetTransportOrderResponseTO setDestinations(List<DestinationState> destinations) {
    this.destinations = requireNonNull(destinations, "destinations");
    return this;
  }

  @Nonnull
  public List<Property> getProperties() {
    return properties;
  }

  public GetTransportOrderResponseTO setProperties(
      @Nonnull
      List<Property> properties
  ) {
    this.properties = requireNonNull(properties, "properties");
    return this;
  }
}
