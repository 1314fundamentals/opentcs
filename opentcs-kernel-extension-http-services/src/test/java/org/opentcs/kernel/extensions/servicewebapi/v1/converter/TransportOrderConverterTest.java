// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi.v1.converter;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.is;

import java.time.Instant;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;
import org.opentcs.data.model.Point;
import org.opentcs.data.model.Vehicle;
import org.opentcs.data.order.DriveOrder;
import org.opentcs.data.order.OrderSequence;
import org.opentcs.data.order.TransportOrder;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.GetTransportOrderResponseTO;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.getevents.OrderStatusMessage;

/**
 * Tests for {@link TransportOrderConverter}.
 */
public class TransportOrderConverterTest {
  private final TransportOrderConverter transportOrderConverter = new TransportOrderConverter();

  @Test
  void checkToOrderStatusMessage() {
    TransportOrder t1
        = new TransportOrder(
            "T1", List.of(
                new DriveOrder(new DriveOrder.Destination(new Point("P1").getReference()))
            )
        )
            .withDispensable(true)
            .withPeripheralReservationToken("some token")
            .withWrappingSequence(new OrderSequence("Os1").getReference())
            .withType("some type")
            .withIntendedVehicle(new Vehicle("V1").getReference())
            .withProcessingVehicle(new Vehicle("V2").getReference())
            .withState(TransportOrder.State.DISPATCHABLE)
            .withProperties(Map.of("some-key", "some-value"));

    OrderStatusMessage statusMessage = transportOrderConverter.toOrderStatusMessage(
        t1, 555, Instant.EPOCH
    );

    assertThat(statusMessage.getOrderName(), is("T1"));
    assertThat(statusMessage.getDestinations().size(), is(1));
    assertThat(statusMessage.getDestinations().getFirst().getLocationName(), is("P1"));
    assertThat(statusMessage.isDispensable(), is(true));
    assertThat(statusMessage.getPeripheralReservationToken(), is("some token"));
    assertThat(statusMessage.getWrappingSequence(), is("Os1"));
    assertThat(statusMessage.getOrderType(), is("some type"));
    assertThat(statusMessage.getIntendedVehicle(), is("V1"));
    assertThat(statusMessage.getProcessingVehicleName(), is("V2"));
    assertThat(statusMessage.getOrderState(), is(OrderStatusMessage.OrderState.DISPATCHABLE));
    assertThat(statusMessage.getProperties().size(), is(1));
    assertThat(statusMessage.getProperties().getFirst().getKey(), is("some-key"));
    assertThat(statusMessage.getProperties().getFirst().getValue(), is("some-value"));
  }

  @Test
  void checkToGetTransportOrderResponse() {
    TransportOrder t1
        = new TransportOrder(
            "T1", List.of(
                new DriveOrder(new DriveOrder.Destination(new Point("P1").getReference()))
            )
        )
            .withDispensable(true)
            .withPeripheralReservationToken("some token")
            .withWrappingSequence(new OrderSequence("Os1").getReference())
            .withType("some type")
            .withIntendedVehicle(new Vehicle("V1").getReference())
            .withProcessingVehicle(new Vehicle("V2").getReference())
            .withState(TransportOrder.State.DISPATCHABLE)
            .withProperties(Map.of("some-key", "some-value"));

    GetTransportOrderResponseTO response = transportOrderConverter.toGetTransportOrderResponse(t1);

    assertThat(response.getName(), is("T1"));
    assertThat(response.getDestinations().size(), is(1));
    assertThat(response.getDestinations().getFirst().getLocationName(), is("P1"));
    assertThat(response.isDispensable(), is(true));
    assertThat(response.getPeripheralReservationToken(), is("some token"));
    assertThat(response.getWrappingSequence(), is("Os1"));
    assertThat(response.getType(), is("some type"));
    assertThat(response.getIntendedVehicle(), is("V1"));
    assertThat(response.getProcessingVehicle(), is("V2"));
    assertThat(response.getState(), is(TransportOrder.State.DISPATCHABLE));
    assertThat(response.getProperties().size(), is(1));
    assertThat(response.getProperties().getFirst().getKey(), is("some-key"));
    assertThat(response.getProperties().getFirst().getValue(), is("some-value"));
  }
}
