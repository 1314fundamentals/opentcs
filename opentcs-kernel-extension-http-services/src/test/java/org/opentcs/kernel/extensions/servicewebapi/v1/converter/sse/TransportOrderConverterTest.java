// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi.v1.converter.sse;

import static org.assertj.core.api.Assertions.assertThat;

import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.List;
import java.util.Map;
import java.util.Set;
import org.approvaltests.Approvals;
import org.approvaltests.core.Options;
import org.assertj.core.groups.Tuple;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.EnumSource;
import org.opentcs.data.ObjectHistory;
import org.opentcs.data.model.Location;
import org.opentcs.data.model.LocationType;
import org.opentcs.data.model.Path;
import org.opentcs.data.model.Point;
import org.opentcs.data.model.Vehicle;
import org.opentcs.data.order.DriveOrder;
import org.opentcs.data.order.OrderSequence;
import org.opentcs.data.order.ReroutingType;
import org.opentcs.data.order.Route;
import org.opentcs.data.order.TransportOrder;
import org.opentcs.kernel.extensions.servicewebapi.JsonBinder;
import org.opentcs.kernel.extensions.servicewebapi.TimestampScrubber;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.sse.ObjectHistoryTO;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.sse.TransportOrderEventTO;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.sse.TransportOrderEventTO.DriveOrderTO;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.sse.TransportOrderEventTO.DriveOrderTO.RouteTO.ReroutingTypeTO;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.sse.TransportOrderEventTO.DriveOrderTO.RouteTO.VehicleOrientationTO;

/**
 * Tests for {@link TransportOrderConverter}.
 */
class TransportOrderConverterTest {

  private JsonBinder jsonBinder;
  private TransportOrderConverter converter;

  @BeforeEach
  void setUp() {
    jsonBinder = new JsonBinder();
    converter = new TransportOrderConverter();
  }

  @Test
  void convert() {
    TransportOrder order = new TransportOrder("order-1", List.of())
        .withProperties(Map.of("key-1", "value-1", "key-2", "value-2"))
        .withHistoryEntry(
            new ObjectHistory.Entry(Instant.EPOCH, "event-code-1", List.of("supplement-1"))
        )
        .withType("type-1")
        .withDependencies(
            Set.of(
                new TransportOrder("dependency-1", List.of()).getReference(),
                new TransportOrder("dependency-2", List.of()).getReference()
            )
        )
        .withPeripheralReservationToken("reservation-token-1")
        .withCurrentDriveOrderIndex(3)
        .withCurrentRouteStepIndex(2)
        .withState(TransportOrder.State.FINISHED)
        .withCreationTime(Instant.EPOCH)
        .withDeadline(Instant.EPOCH)
        .withFinishedTime(Instant.EPOCH)
        .withIntendedVehicle(new Vehicle("vehicle-1").getReference())
        .withProcessingVehicle(new Vehicle("vehicle-2").getReference())
        .withWrappingSequence(new OrderSequence("sequence-1").getReference())
        .withDispensable(true);
    order = order.withDriveOrders(
        List.of(
            new DriveOrder(
                "drive-order-1",
                new DriveOrder.Destination(
                    new Location(
                        "location-1",
                        new LocationType("location-type-1").getReference()
                    ).getReference()
                )
                    .withOperation("operation-1")
                    .withProperties(Map.of("key-3", "value-3"))
            )
                .withTransportOrder(order.getReference())
                .withRoute(
                    new Route(
                        List.of(
                            new Route.Step(
                                new Path(
                                    "path-1-2",
                                    new Point("point-1").getReference(),
                                    new Point("point-2").getReference()
                                ),
                                new Point("point-1"),
                                new Point("point-2"),
                                Vehicle.Orientation.FORWARD,
                                1,
                                200
                            ),
                            new Route.Step(
                                null,
                                null,
                                new Point("point-2"),
                                Vehicle.Orientation.FORWARD,
                                2,
                                100
                            )
                                .withReroutingType(ReroutingType.REGULAR)
                                .withExecutionAllowed(false)
                        )
                    )
                )
                .withState(DriveOrder.State.FINISHED)
        )
    );

    TransportOrderEventTO.TransportOrderTO result = converter.convert(order);

    Approvals.verify(
        jsonBinder.toJson(result),
        // Scrub timestamps in the result's JSON representation as some of them may be generated
        // during test runtime, making it impossible to use ApprovalTests.
        new Options(new TimestampScrubber())
    );
  }

  /**
   * Ensures that timestamps are converted correctly. (Testing this is necessary, since timestamps
   * are scrubbed in the ApprovalTest and therefore their conversion is not covered there.)
   */
  @Test
  void convertsTimestampsCorrectly() {
    Instant time = Instant.parse("2025-07-24T12:03:00Z");
    TransportOrder order = new TransportOrder("order-1", List.of())
        .withCreationTime(time)
        .withDeadline(time.plus(1, ChronoUnit.HOURS))
        .withFinishedTime(time.plus(2, ChronoUnit.HOURS))
        .withHistoryEntry(
            new ObjectHistory.Entry(
                time.plus(3, ChronoUnit.HOURS),
                "event-code-1",
                List.of("supplement-1")
            )
        );

    TransportOrderEventTO.TransportOrderTO result = converter.convert(order);

    assertThat(result.getCreationTime()).isEqualTo(time);
    assertThat(result.getDeadline()).isEqualTo(time.plus(1, ChronoUnit.HOURS));
    assertThat(result.getFinishedTime()).isEqualTo(time.plus(2, ChronoUnit.HOURS));
    assertThat(result.getHistory().getEntries())
        .extracting(
            ObjectHistoryTO.ObjectHistoryEntryTO::getTimestamp,
            ObjectHistoryTO.ObjectHistoryEntryTO::getEventCode,
            ObjectHistoryTO.ObjectHistoryEntryTO::getSupplements
        )
        .contains(
            Tuple.tuple(
                time.plus(3, ChronoUnit.HOURS),
                "event-code-1",
                List.of("supplement-1")
            )
        );
  }

  @ParameterizedTest
  @EnumSource(TransportOrder.State.class)
  void convertsAllTransportOrderStates(TransportOrder.State state) {
    TransportOrder order = new TransportOrder("order-1", List.of()).withState(state);

    TransportOrderEventTO.TransportOrderTO result = converter.convert(order);

    TransportOrderEventTO.StateTO expectedState = switch (state) {
      case RAW -> TransportOrderEventTO.StateTO.RAW;
      case ACTIVE -> TransportOrderEventTO.StateTO.ACTIVE;
      case DISPATCHABLE -> TransportOrderEventTO.StateTO.DISPATCHABLE;
      case BEING_PROCESSED -> TransportOrderEventTO.StateTO.BEING_PROCESSED;
      case WITHDRAWN -> TransportOrderEventTO.StateTO.WITHDRAWN;
      case FINISHED -> TransportOrderEventTO.StateTO.FINISHED;
      case FAILED -> TransportOrderEventTO.StateTO.FAILED;
      case UNROUTABLE -> TransportOrderEventTO.StateTO.UNROUTABLE;
    };
    assertThat(result.getState()).isEqualTo(expectedState);
  }

  @ParameterizedTest
  @EnumSource(DriveOrder.State.class)
  void convertsAllDriveOrderStates(DriveOrder.State state) {
    TransportOrder order = new TransportOrder(
        "order-1",
        List.of(
            new DriveOrder(
                "drive-order-1",
                new DriveOrder.Destination(new Point("point-1").getReference())
            ).withState(state)
        )
    );

    TransportOrderEventTO.TransportOrderTO result = converter.convert(order);

    DriveOrderTO.StateTO expectedState = switch (state) {
      case PRISTINE -> DriveOrderTO.StateTO.PRISTINE;
      case TRAVELLING -> DriveOrderTO.StateTO.TRAVELLING;
      case OPERATING -> DriveOrderTO.StateTO.OPERATING;
      case FINISHED -> DriveOrderTO.StateTO.FINISHED;
      case FAILED -> DriveOrderTO.StateTO.FAILED;
    };
    assertThat(result.getDriveOrders().getFirst().getState()).isEqualTo(expectedState);
  }

  @ParameterizedTest
  @EnumSource(Vehicle.Orientation.class)
  void convertsAllVehicleOrientations(Vehicle.Orientation orientation) {
    TransportOrder order = new TransportOrder(
        "order-1",
        List.of(
            new DriveOrder(
                "drive-order-1",
                new DriveOrder.Destination(new Point("point-1").getReference())
            ).withRoute(
                new Route(
                    List.of(
                        new Route.Step(
                            null,
                            null,
                            new Point("point-1"),
                            orientation,
                            0,
                            1
                        )
                    )
                )
            )
        )
    );

    TransportOrderEventTO.TransportOrderTO result = converter.convert(order);

    VehicleOrientationTO expectedOrientation
        = switch (orientation) {
          case FORWARD -> VehicleOrientationTO.FORWARD;
          case BACKWARD -> VehicleOrientationTO.BACKWARD;
          case UNDEFINED -> VehicleOrientationTO.UNDEFINED;
        };
    assertThat(
        result.getDriveOrders().getFirst().getRoute().getSteps().getFirst().getVehicleOrientation()
    ).isEqualTo(expectedOrientation);
  }

  @ParameterizedTest
  @EnumSource(ReroutingType.class)
  void convertsAllReroutingTypes(ReroutingType reroutingType) {
    TransportOrder order = new TransportOrder(
        "order-1",
        List.of(
            new DriveOrder(
                "drive-order-1",
                new DriveOrder.Destination(new Point("point-1").getReference())
            ).withRoute(
                new Route(
                    List.of(
                        new Route.Step(
                            null,
                            null,
                            new Point("point-1"),
                            Vehicle.Orientation.FORWARD,
                            0,
                            1
                        )
                            .withReroutingType(reroutingType)
                    )
                )
            )
        )
    );

    TransportOrderEventTO.TransportOrderTO result = converter.convert(order);

    ReroutingTypeTO expectedReroutingType = switch (reroutingType) {
      case REGULAR -> ReroutingTypeTO.REGULAR;
      case FORCED -> ReroutingTypeTO.FORCED;
    };
    assertThat(
        result.getDriveOrders().getFirst().getRoute().getSteps().getFirst().getReroutingType()
    ).isEqualTo(expectedReroutingType);
  }
}
