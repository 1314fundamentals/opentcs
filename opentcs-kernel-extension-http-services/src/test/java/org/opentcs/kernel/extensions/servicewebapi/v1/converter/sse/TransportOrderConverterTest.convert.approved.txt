{
  "name" : "order-1",
  "properties" : {
    "key-1" : "value-1",
    "key-2" : "value-2"
  },
  "history" : {
    "entries" : [ {
      "timestamp" : "[Timestamp]",
      "eventCode" : "tcs:history:orderCreated",
      "supplements" : [ ]
    }, {
      "timestamp" : "[Timestamp]",
      "eventCode" : "event-code-1",
      "supplements" : [ "supplement-1" ]
    }, {
      "timestamp" : "[Timestamp]",
      "eventCode" : "tcs:history:orderReachedFinalState",
      "supplements" : [ ]
    }, {
      "timestamp" : "[Timestamp]",
      "eventCode" : "tcs:history:orderProcVehicleChanged",
      "supplements" : [ "vehicle-2" ]
    } ]
  },
  "type" : "type-1",
  "dependencies" : [ "dependency-1", "dependency-2" ],
  "driveOrders" : [ {
    "name" : "drive-order-1",
    "destination" : {
      "destination" : "location-1",
      "operation" : "operation-1",
      "properties" : {
        "key-3" : "value-3"
      }
    },
    "transportOrder" : "order-1",
    "route" : {
      "steps" : [ {
        "path" : "path-1-2",
        "sourcePoint" : "point-1",
        "destinationPoint" : "point-2",
        "vehicleOrientation" : "FORWARD",
        "routeIndex" : 1,
        "costs" : 200,
        "executionAllowed" : true,
        "reroutingType" : null
      }, {
        "path" : null,
        "sourcePoint" : null,
        "destinationPoint" : "point-2",
        "vehicleOrientation" : "FORWARD",
        "routeIndex" : 2,
        "costs" : 100,
        "executionAllowed" : false,
        "reroutingType" : "REGULAR"
      } ],
      "costs" : 300
    },
    "state" : "FINISHED"
  } ],
  "peripheralReservationToken" : "reservation-token-1",
  "currentDriveOrderIndex" : 3,
  "currentRouteStepIndex" : 2,
  "state" : "FINISHED",
  "creationTime" : "[Timestamp]",
  "deadline" : "[Timestamp]",
  "finishedTime" : "[Timestamp]",
  "intendedVehicle" : "vehicle-1",
  "processingVehicle" : "vehicle-2",
  "wrappingSequence" : "sequence-1",
  "dispensable" : true
}